pennylane/ops/mid_circuit_measure.py

```diff
@@ -0,0 +1,291 @@
+"""
+Mid-circuit measurements and associated operations.
+"""
+
+import functools
+
+from typing import Union, Any, TypeVar, Generic, Callable, Type
+
+from pennylane.operation import Operation
+
+
+def mid_measure(wire):
+    """
+    Create a mid-circuit measurement and return an outcome.
+
+    .. code-block:: python
+
+        m0 = qml.Measure(0)
+    """
+    wire_id = wire
+    _MidCircuitMeasure(wire_id, wire)
+    return MeasurementDependantValue(wire_id, 0, 1)
+
+
+class _MidCircuitMeasure(Operation):
+    """
+    Operation to perform mid-circuit measurement.
+    """
+
+    num_wires = 1
+
+    def __init__(self, measure_var, wires=None):
+        self.measure_var = measure_var
+        self.runtime_value = None
+        super().__init__(wires=wires)
+
+
+# pylint: disable=protected-access
+def apply_to_measurement_dependant_values(fun):
+    """
+    Apply an arbitrary function to a `MeasurementDependantValue` or set of `MeasurementDependantValue`s.
+
+    Ex:
+
+    .. code-block:: python
+
+        m0 = qml.mid_measure(0)
+        m0_sin = qml.apply_to_measurement_dependant_values(np.sin)(m0)
+    """
+
+    @functools.wraps(fun)
+    def wrapper(*args, **kwargs):
+        partial = _Value()
+        for arg in args:
+            partial = partial._merge(arg)
+        partial._transform_leaves_inplace(
+            lambda *unwrapped: fun(*unwrapped, **kwargs)  # pylint: disable=unnecessary-lambda
+        )
+        return partial
+
+    return wrapper
+
+
+T = TypeVar("T")
+
+
+# pylint: disable=protected-access
+class MeasurementDependantValue(Generic[T]):
```

Comment:
Is there a clear path for users to return the expectation value of mid-circuit measurement results?
E.g.
```python
import pennylane as qml

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def f(x):
    qml.RX(x, wires=0)
    qml.RY(x ** 2, wires=1)
    qml.CNOT(wires=[0, 1])
    qml.RX(x, wires=1)

    m = qml.mid_measure(1)
    qml.condition(qml.RY)(m, wires=0)
    return qml.expval(qml.PauliZ(0)), qml.expval(m)

f(0.4)
```